src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,35 +121,176 @@ def __init__(
+                except Exception:
+                    logger.error(
+                        "Async future failed in billing metrics consumer.",
+                        exc_info=future.exception(),
+                        extra={"offset": metrics_msg.offset},
+                    )
+                    # Don't commit the future when it fails -- if no futher messages
+                    # are committed before shutting down the consumer, new consumers
+                    # will process the current offset again.
+                    do_commit = False
+                    self._ongoing_billing_outcomes.popleft()
+
+            if do_commit:
+                self._mark_commit_ready()
+                self._bulk_commit()
+
+        self._billing_producer.close()
+
     def terminate(self) -> None:
         self.close()
+        amount_dropped = len(self._ongoing_billing_outcomes)
+        while self._ongoing_billing_outcomes:
+            ongoing_work, _ = self._ongoing_billing_outcomes.popleft()
+            if ongoing_work:
+                ongoing_work.cancel()
+        amount_dropped += self._clear_ready_queue()
+        if amount_dropped > 0:
+            logger.warning(f"terminated, items dropped: {amount_dropped}")
+
+        self._billing_producer.close()
 
     def close(self) -> None:
-        self.__closed = True
+        self._closed = True
 
     def submit(self, message: Message[KafkaPayload]) -> None:
-        assert not self.__closed
-        self.__messages_since_last_commit += 1
+        assert not self._closed
 
-        payload = self._get_payload(message)
-        self._produce_billing_outcomes(payload)
-        self._mark_commit_ready(message)
+        bucket_payload = self._get_bucket_payload(message)
+        quantity = self._count_processed_transactions(bucket_payload)
 
-    def _get_payload(self, message: Message[KafkaPayload]) -> MetricsBucket:
+        if quantity < 1:
+            # We still want to commmit buckets that don't generate billing
+            # outcomes, since the consumer has already processed them. To keep
+            # the offset order when committing, we still need to go through the
+            # futures queue.
+            future = None
+        else:
+            value = json.dumps(
+                {
+                    "timestamp": datetime.now(),
+                    "org_id": bucket_payload["org_id"],
+                    "project_id": bucket_payload["project_id"],
+                    "outcome": Outcome.ACCEPTED.value,
+                    "category": DataCategory.TRANSACTION,
+                    "quantity": quantity,
+                }
+            ).encode("utf-8")
+            billing_payload = KafkaPayload(key=None, value=value, headers=[])
+            future = self._billing_producer.produce(
+                destination=self._billing_topic, payload=billing_payload
+            )
+
+        self._ongoing_billing_outcomes.append(MetricsBucketBilling(future, message))
```

Comment:
Wouldn't it be easier to just not append to this list when future = None?

Reply:
The consumer should commit the buckets that don't produce anything, yet it has already processed these messages. If the consumer doesn't append anything on `future = None` cases, it won't commit any messages until it processes a message that does produce an outcome. Do you have a suggestion on a better way to solve this use case?

Reply:
No. I did not realise you were skipping so many where it would actually matter. This is fine then.